core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicPath.java

```diff
@@ -0,0 +1,164 @@
+    private final String skeleton;
+
+    /**
+     * Set of the variables declared in this dynamic path.
+     */
+    private final ImmutableSet<String> variables;
+
+    DynamicPath(String str) {
+        StringJoiner patternJoiner = new StringJoiner("/");
+        StringJoiner skeletonJoiner = new StringJoiner("/");
+        Set<String> placeholders = new HashSet<>();
+        for (String token : str.split("/")) {
+            String variable = variable(token);
+            if (variable != null) {
+                if (!placeholders.contains(variable)) {
+                    /**
+                     * If given token is a first occurred variable mark, add it to the variable set and replace
+                     * it to named group expression in regex.
+                     */
+                    placeholders.add(variable);
+                    patternJoiner.add(String.format("(?<%s>[^/]+)", variable));
+                } else {
+                    /**
+                     * If given token is a re-occurred variable mark, replace it to backreference expression
+                     * in regex.
+                     */
+                    patternJoiner.add(String.format("\\k<%s>", variable));
+                }
+                skeletonJoiner.add("{}");
+            } else {
+                /**
+                 * if given token is a constant, does not manipulate it.
+                 */
+                patternJoiner.add(token);
+                skeletonJoiner.add(token);
+            }
+        }
+
+        pattern = Pattern.compile(patternJoiner.toString());
+        skeleton = skeletonJoiner.toString();
+        variables = ImmutableSet.copyOf(placeholders);
+
+        if (variables.isEmpty()) {
+            throw new IllegalArgumentException("No placeholder exists in given path: " + str);
+        }
+    }
+
+    /**
+     * Returns variable contained in the path element. If it contains no variable, returns null.
+     *
+     * <p>e.g. {variable} -> variable
+     * :variable -> variable
+     * constant -> constant
+     */
+    private String variable(String token) {
+        if (token.startsWith("{") && token.endsWith("}")) {
+            return token.substring(1, token.length() - 1);
+        }
+
+        if (token.startsWith(":")) {
+            return token.substring(1);
+        }
+
+        return null;
+    }
+
+    /**
+     * Returns the skeleton.
+     */
+    public String skeleton() {
+        return skeleton;
+    }
+
+    /**
+     * Returns the variables.
+     */
+    public Set<String> variables() {
+        return variables;
+    }
+
+    /**
+     * Returns Binding results with given path. If given path does not match, returns null.
+     */
+    public Map<String, String> bind(String path) {
+        Matcher matcher = pattern.matcher(path);
+        if (!matcher.matches()) {
+            return null;
+        }
+
+        ImmutableMap.Builder<String, String> builder = ImmutableMap.<String, String>builder();
```

Comment:
I considered using the stream for the first time but changed my mind not to using it because `DynamicPath#bind` is called so frequently.